CenterForOpenScience/osf.io
# [ENG-637] Feature/sparse lists

## Purpose

Regular list endpoints are slow because they serialize more than is needed for most cases. However, ember requires consistency between requests for items of the same combination of `id` and `type`. This attempts to solve that by providing a new set of endpoints for Nodes and Registrations that uses sparse serialization and provides a new `type` for the objects, thus keeping Ember from being confused.

## Changes

1. New sparse-[node, registration]` endpoints with all the fixins
2. Tests

## QA Notes

Testing for this is purely through the API. There are a new set of endpoints under `/v2/sparse/` that get nodes and registrations. List of endpoints below. When you use these endpoints, 1) they should be much faster; 2) you should get a limited subset of the fields. According to @jamescdavis:

The fields/relationships the FE needs for node lists are:

* title
* description
* date_created
* date_modified
* public
* fork
* category
* currentUserPermissions
* contributors
* tags

Registrations also need:

* pendingRegistrationApproval
* withdrawn
* pendingWithdrawal
* pendingEmbargoTerminationApproval
* embargoed
* pendingEmbargoApproval
* archiving
* registrationSchema
* registeredMeta

There will be extra fields there because we need them; the above list is just what was explicitly requested by Front End. As mentioned above, these lists should be no slower and preferably much faster than the non-sparse versions of the lists. If relationships are to a node or a registration, it should reference the sparse version of that (except for the `detail` relationship, which should be the full version). Self link should go to the sparse detail link.

The specific endpoints added under `/v2/sparse/` are:
1. nodes
2. nodes/<node_id>/
3. nodes/<node_id>/children
4. nodes/<node_id>/linked_nodes
5. nodes/<node_id>/linked_registrations
6. registrations
7. registrations/<registration_id>/
8. registrations/<registration_id>/children
9. registrations/<registration_id>/linked_nodes
10. registrations/<registration_id>/linked_registrations
11. users/<user_id>/nodes
12. users/<user_id>/registrations
	
## Documentation

We are going to keep this undocumented for now. It may eventually be documented, but we want to maintain the flexibility of undocumented endpoints until the front end is happy with the functionality, as this is a brand-new paradigm for api serialization.

## Side Effects

This should not have any effect on existing functionality.

## Ticket

https://openscience.atlassian.net/browse/ENG-637
